### ISSUE TITLE
GNOME 45 support

### ISSUE THREAD

**Issue by yochananmarqos** [gTile/gTile#336]
There are breaking changes which will require a separate release for 45 and <=44:

See https://gjs.guide/extensions/upgrading/gnome-shell-45.html

**Comment by scherepanov** [comment]
Great news!
At least, all old versions <=44 will be on stable code version :-) no upgrades for them.
And all code that branching based on gnome-shell version now can be removed.

When gnome-shell 45 will be available?
Can you submit PR with changes to make gTile to run on v45? Would greatly appreciate our efforts.

**Comment by yochananmarqos** [comment]
[GNOME 45.beta](https://discourse.gnome.org/t/gnome-45-beta-released/16766) is currently available. See the [GNOME 45 Schedule](https://wiki.gnome.org/FortyFive) for more info.

This was just a heads up. I'm only a package maintainer, not a developer.

**Comment by scherepanov** [comment]
Thanks!

**Comment by GerryT11** [comment]
FYI, here is some more information on the changes in Gnome 45 for extensions: https://blogs.gnome.org/shell-dev/2023/09/02/extensions-in-gnome-45/ Thanks for gTile!

**Comment by okoponen** [comment]
Have been using gTile for ~3 years and basically, can't live without it.

Tried to convert this to Gnome45, but seems that TypeScript is not able to resolve the  new import paths e.g.
```
import GLib from 'gi://GLib';
```

For the most parts, these can be circumvented by 
```
// @ts-ignore
```

but for any class, that extends anything exported, e.g. extending the class app 
```
export default class App extends Extesion
```
or
```
const GTileStatusButton = GObject.registerClass({
    GTypeName: 'GTileStatusButton',
}, class GTileStatusButton extends PanelMenu.Button {
```
the extended class typing should be needed  to get the super-methods, e.g. in the latter example, this.add_actor()

I suspect, that dropping the TS support would make migration to Gnome45 easier. That said, the benefits of TS are great and falling back to plain JS is something I would encourage.

Not sure, what options there are, other than trying to manually fix the remaining things on JS-converted code directly or somehow fix the imports for GJS. Seems that most GJS examples area still heavily using the "imports" pragma.

**Comment by okoponen** [comment]
I encourage TS usage. Use it in work and love it.

I have presented this dilemma to Gnome Discourse and tagged a previous TS in GJS evangelist to the thread. Fingers crossed...

**Comment by pavangayakwad** [comment]
I am on Gnome 45+ (OpenSUSE) and looking forward to this fix desperately!

**Comment by scherepanov** [comment]
@okoponen any words from GJS evangelist?

**Comment by okoponen** [comment]
@scherepanov 

Zero answers to my post. There was another similar question, which did not get that much attention either...

Not sure if the easiest way to solve would be to build it as JS, then spam ts-ignore like a crazy person.

**Comment by renataogarcia** [comment]
@okoponen 

Maybe use paths in TS config? Something along these lines:

```json 
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "gi:*": ["path/to/match/*"]
    }
  }
}
```

**Comment by HeavenVolkoff** [comment]
Maybe https://github.com/gjsify/ts-for-gir can help?

**Comment by okoponen** [comment]
> Maybe https://github.com/gjsify/ts-for-gir can help?
> 
> It has pre-generated gobject types for typescript available on NPM: https://github.com/gjsify/types https://www.npmjs.com/package/@girs/glib-2.0

Yes, these do help to the 'gi://' typed imports. So far, generating these myself seems to work, but managed to (quickly) generate only part of those imports (Shell, Gtk). Using the pre-defined packages did not work as per documentation and did not yet  find sufficient time to troubleshoot, why those did not work.

AFAIK, this package will still leave the 'resouce:///org.gnome.shell.....' styled imports unsolved.

**Comment by mfixman** [comment]
Is there any temporary patch I can apply to make gTile compatible with GNOME 45?

**Comment by scherepanov** [comment]
It appears, we do not have any workaround. Currently it is not possible to build gTile for GNOME 45.

**Comment by sardaukar** [comment]
That's terrible news, gTile is one of my "must have" extensions to cope with Gnome's default window placement logic :(

**Comment by HeavenVolkoff** [comment]
@okoponen I think the `'resouce:///org.gnome.shell.....'` imports can be solved by adpating this project: https://github.com/gjsify/gnome-shell/tree/main/packages/gnome-shell (from the same creator of [ts-for-gir](https://github.com/gjsify/ts-for-gir)). Unfortunately, the package can't be used directly because the glue code present in the `.ts` files is using the old `import` syntax. But maybe it works if you copy the required typing definitions from https://github.com/gjsify/gnome-shell/tree/main/packages/gnome-shell/src in the expected hierarchy and instruct the ts-compiler to alias the 'resouce:' import using the `paths` config:
https://www.typescriptlang.org/tsconfig/#paths

```
{
  "compilerOptions": {
    "paths": {
      "resource:///org/gnome/shell/*": ["./@types/gnome/*"]
    }
  }
}
```

**Comment by schnz** [comment]
> I think quite a few gnome-shell extensions are using TS, and facing exactly same problem. Will watch what solution will popup in a wild.

I've just read through this issue and did a bit of research on the subject. Although I've never written or contributed to a Gnome extension, I would probably refer to https://gitlab.gnome.org/BrainBlasted/gnome-typescript-template/ as a starting point, which is referenced [here](https://release.gnome.org/45/developers/)

---

Is anyone currently actively working on a GNOME 45 compatible release? I may find a day or two next week to look into it and would use the gnome typescript template as a starting point. But as I said, its new territory for me - so don't get your hopes up too high.

**Comment by okoponen** [comment]
@schnz 

I did start the migrate, but not currently working actively with it. I made a fork
https://github.com/okoponen/gTile-gnome45-wip

Specifically, the problems I encountered, what I have done, can be found in
https://github.com/okoponen/gTile-gnome45-wip/blob/main/GNOME45-migration.txt

I am happy to help, but not sure how much help I am, having 0 knowledge of Gnome developing is surely not helping...

**Comment by schnz** [comment]
Preliminary assessment:
* I've managed to get the type completion working for both `gi://*` and `resource://*` imports.
* I've not yet managed to compile the project due to lots of (seemingly legit) errors (118 errors, to be exact)
* I could not (yet?) find all GNOME-45 equivalent imports for everything:
  * `imports.signals` seems to be no longer available. IMO this requires refactoring of all classes that are currently passed to `Signals.addSignalMethods` by extending them from `GObject.Object` (see also [Migrating Shell Classes](https://wiki.gnome.org/Projects/GnomeShell/Extensions/MigratingShellClasses))
  * `imports.mainloop` seems to have no equivalent. It seems to be of use for one purpose only, which is [here](https://github.com/gTile/gTile/blob/f19decf6b4296d89158303c5cb5e49d953d2896f/app.ts#L701). Maybe someone more familiar with the extension can explain if/why this is needed and if its possible to implement differently.

I will give this another day of work and see where this goes. Hopefully at the very least someone else find some use in what I did. Maybe I even get it running. Lets see.

**Comment by HeavenVolkoff** [comment]
@schnz `imports.signals` is not exported to ESM. I am not sure if it is deprecated, but there is a notice in gjs docs:
https://gitlab.gnome.org/GNOME/gjs/-/blob/master/doc/Signals.md#import

But as you pointed out, we can replace that behavior with `GObject.registerClass`:
(WebArchive link because https://gjs.guide was down at the time when I was writing this)
https://web.archive.org/web/20230529160535/https://gjs.guide/guides/gobject/subclassing.html#subclassing-gobject

About the [`MainLoop` usage](https://gitlab.gnome.org/GNOME/gjs/-/blob/master/doc/Mainloop.md#mainlooptimeout_addtimeout-handler-priority), I think it should be replaced with a [`setInterval`](https://gitlab.gnome.org/GNOME/gjs/-/blob/master/doc/Timers.md#setintervalhandler-timeout-arguments) now that gjs supports it, and backward compatibility is being broken anyway.

I want to help with the porting, but I am very busy at work rn. However, whenever I get some free time and if this is not solved yet, I intend to attempt porting gTitle to gnome 45, because it is an exceptionally useful extension that has been a part of my workflow for some time, and I am missing it quite a lot

**Comment by schnz** [comment]
Quick update on the situation:

**TL;DR** I am working on a rewrite of the extension.

**Rationale**
My original approach was to adapt the existing code base for compatibility with Gnome 45 (see PR #340). The plan was to get it up & running first with replaced imports and no errors at compile time and then step by step refactor the remaining code until the extension is fully functional. However, even though I made a good progress in getting a compile-time error-free project, I wasn't really actually be able to compile it with the Bazel toolchain that gTile uses. I spent a fair amount of time in learning the Basics of Bazel and its rules-system but at the end of the day I was only fighting with the rules that were supposed to ease the experience in building the project (but failed to to do so, mainly because it relied on older TypeScript and pnpm versions than the current stable ones).

So instead, I tried to setup a minimal TypeScript project for a Gnome extension from scratch and the plan was then to copy & paste the relevant parts from gTile over to the new project. But there is some refactoring required to get the project running due to deprecation of APIs that are no longer available in GJS 1.78 (Gnome 45). So I somehow ended up simply porting gTile step by step and rewrite all the code that is either completely deprecated (and *requires* refactoring) or which idioms are no longer recommended to be used (and thus it *should* be refactored). 

The good news is, that there will be a Gnome 45 compatible gTile version in hopefully a week or so. The code base will be (IMO) much more easily be maintainable and also relies solely on npm (and tsc) as a build tool (at least for now). The bad news is that I don't know how this PR will be received by the project maintainers who will have to deal with a complete new code base in terms of its directory structure, module splitting and build system.

**Comment by ritchie888** [comment]
Much appreciated, schnz (and anyone else working on this port)!

**Comment by scherepanov** [comment]
@schnz great news!

gTile would benefit from full re-write!

Instead of PR, create a new branch. I suggest to add you to admins of gTile, let me know.

We should try to make new gTile compatible with pre-45, but not necessary if it is too much work.

All parts of current gTile have satisfying functionality and can be ported to rewrite (of course with better refactoring). Exception is autotiling - it has to be deeply redesigned.

Autotiling is very tricky - there is a huge demand form users, but extreme variety of autotiling. gTile implement only few autotilings and it fits demands of very small fraction of users. (My estimate of gTile user base install is 35 000 - 50 000, based on downloads of new versions)

We need to completely rewrite autotiling. I had an idea for a long time how to satisfy all users demands for various autotiling. Generally speaking, I want to have plugin with code editable straight in preferences and saved in config. gTile will load plugins code from config on start, and will execute in sandbox. We will pass to plugin object with all required info for autotiling - list of windows, their sizes/position etc. Plugin will be responding with list of windows to resize and their new sizes/positions. Grand idea is to let users to edit plugin code on a fly, save to config and reload plugin, and have new code running. Plugin will be executing in sandbox, dynamic code load will not be a problem. gTile will provide keyboard binding for plugin invocation. Preferences will have per-plugin config. Goal is to let users to create autotiling, publish it (for starting to gTile repo), and all users will be able to use it. 

Good goal would be to implement i3-like windowing control with plugin. And so on, there is unlimited number of possible autotiling that I can tell right now.

I am looking forward to make gTile more versatile, providing best gnome-shell autotiling capabilities.

**Comment by schnz** [comment]
As it turns out, the preference window also requires a rewrite due to a breaking API change. I just now realized that gTile happens to have what is called _global hotkeys_ that are supposed to work even when the extension is disabled. I find that feature counterintuitive, to say the least. Is there any (valid) reason to keep them? Because I am inclined to remove this feature and make hotkeys only work when the extension is active.

**Comment by scherepanov** [comment]
Yes, gTile packs quite a lot of functionality. Will take some time to implement it all.

**Comment by schnz** [comment]
> Instead of PR, create a new branch. I suggest to add you to admins of gTile, let me know.

I take you up on that offer. There is still plenty of work to do, but it reaches a point where the code base becomes more and more stable. It needs a final iteration until it has a state that is worthy of being published. As of now it still lacks most of the features. But the ground work is laid and implementing features will be nowhere as time consuming once the basic architecture is completed.

By the way: I plan to release the rewrite of the extension under the MIT license, i.e., unless someone raises concerns that cause me to reconsider that decision. Its more permissive than the current GPL 2 license.

Since this is a complete rewrite there __will__ be bugs. There will probably also be some breaking changes with regards to the behavior of gTile. Either by accident, e.g, because of a different set of APIs being used, or on purpose because I felt it would improve the user experience. This means that I would really like to have some kind of alpha/beta phase where users can report bugs and inconveniences before the first version is published on extensions.gnome.org. This will hopefully also help to find bugs that appear in distros other than Arch Linux (which is the only distro that I am testing it on). So when the time comes, I hope that some of you will volunteer to beta test the extension.

**Comment by scherepanov** [comment]
@schnz sent you invitation.
Originally I created this repo with gpl3, only to find that to publish extension on EGO you have to certify it is GPL v2+ (kinda LGPL2). MIT license would not do.

**Comment by schnz** [comment]
Thanks for the invite!

> Originally I created this repo with gpl3, only to find that to publish extension on EGO you have to certify it is GPL v2+ (kinda LGPL2). MIT license would not do.

Good point. I wasn't aware of that. I suppose dual-licensing would be an option. But since the most likely use of the code would be as part of an extension and all extensions have to be GPLv2, I will go with that, too.

**Comment by kamil-karkus** [comment]
> So when the time comes, I hope that some of you will volunteer to beta test the extension

I can test this on multiple Ubuntu 23.10 (wayland & xorg) instances.

**Comment by schnz** [comment]
@kamil-karkus : I have to wonder: Was gTile supposed to be working on Ubuntu in the first place? Because I am using a Ubuntu 23.04 live USB distro to see how gTile acts in certain situations and I it seems almost non-functional there. For example, most of the times I try to select a tile area in the grid, the mouse clicks are simply targeting the application below the gTile overlay.

**Comment by lckarssen** [comment]
@schnz I am currently using gTile on Ubuntu 23.04 (with Wayland). I haven't encountered any problems yet, but then again, I'm not really using the mouse with gTile, I only use keyboard shortcuts. 

I did a few quick tests with the mouse and I didn't notice any weird effects. However, this might have to do with the fact that I use 'focus follows mouse' mode (sloppy focus), so when I use gTile, the application window I want to tile is always already selected.

**Comment by kamil-karkus** [comment]
@schnz I use only keyboard shortcuts, but I've just did few quick tests with the mouse and it seems to just work fine on 23.04.

**Comment by okoponen** [comment]
@schnz I have used gTile from Ubuntu 20.04 - 23.04. In the early versions, i did use xorg, but i think I made the switch to Wayland quite early already on 20.04. I have used mainly keyboard shortcuts, but occasionally (couple times a year) when i have used the super + enter / grid with mouse, it did work. Can't say for sure in 23.04.

I can also help test on Ubuntu 23.10, at least the keyboard shortcuts. I noticed, that gTile supports several different ways of working, mine was quite close to what ubuntu-tiling-assistant delivers. That said, there are still some cases, where I wish really badly, that I would have gTile instead.

**Comment by schnz** [comment]
It's funny. I usually worked with the default Gnome keybindings to move and resize my windows (`<Alt>+F7`, `<Alt>+F8`, `<Super>+<Arrow Keys>`). I only ever saw use in gTile with the `<Super>+<Enter>` shortcut for tiling my windows manually with the mouse 🤣.  Only now that I rewrote the entire thing I realize that it also provides some handy shortcuts.

**Comment by schnz** [comment]
🥳 I just [pushed](https://github.com/gTile/gTile/tree/next) the current state of development and created a new PR (#341) which I will keep updating in the coming days. It's still in an early stage (also way too early for any alpha/beta testing). But whoever is interested can check it out. Installation instructions are provided in #341.

Although it lacks almost all features, it is a pretty stable and clean code base. So anyone interested in participating or helping to boost the development process: please say the word. Especially the preference dialog could be developed pretty much without interference. Although, before you do, please contact me first. I have some changes in mind for it.

**Comment by SoiledBrush** [comment]
Just tried it, works on fedora 39 beta👍. Auto close on resize doesn't work. 
Can't wait for keyboard shortcuts.

**Comment by ritchie888** [comment]
I have tried the new version, and it's great! Similarly to @SoiledBrush, I'm also looking forward to keyboard shortcuts as that's how I almost exclusively use the extension. I can also confirm that auto-close doesn't work (Arch Linux). Great work!

**Comment by hengestone** [comment]
@schnz Thanks for working on this! As a token of appreciation, I'd like you buy you a coffee. Or something :)
If possible, could you sign up for getting sponsored on github?

**Comment by yochananmarqos** [comment]
@schnz: I tried building from the next branch, but it can't get past `npm install`:

```
npm ERR! code E401
npm ERR! 401 Unauthorized - GET https://npm.pkg.github.com/download/@schnz/xlib-2.0/2.0.0-3.2.2/59c387145324f43003b95364a79fb69bef143443 - authentication token not provided

npm ERR! A complete log of this run can be found in: /build/gnome-shell-extension-gtile-git/src/npm_cache/_logs/2023-11-03T23_12_54_231Z-debug-0.log
```

I can't find anything on my end that wold be causing it.

**Comment by schnz** [comment]
> @schnz Thanks for working on this! As a token of appreciation, I'd like you buy you a coffee. Or something :)
If possible, could you sign up for getting sponsored on github?

That’s very kind! Much appreciated. I've submitted my GitHub sponsorship profile which now is pending for approval.

> @schnz: I tried building from the next branch, but it can't get past npm install:
Have you checked the installation instructions in #341?

By the way: I am just preparing the next commit which brings additional functionality back. It will be ready within the hour. Although hotkeys are still missing, sorry to disappoint.

**Comment by yochananmarqos** [comment]
> Have you checked the installation instructions in #341?

@schnz  Ah, yes. Well, that's not acceptable for reproducible builds. Neither my AUR package nor the Arch extra package will be updated until it can be built without authentication.

**Comment by schnz** [comment]
> @schnz Ah, yes. Well, that's not acceptable for reproducible builds. Neither my AUR package nor the Arch extra package will be updated until it can be built without authentication.

It doesn't prevent reproducible builds. But I am aware of the fact that this won't be a viable solution for a production release. I'd hoped that the owner of the [gjsify](https://github.com/gjsify) project would be interested in providing an up-to-date NPM package. I've already [reached out](https://github.com/gjsify/gnome-shell/issues/5) to no avail.

If by the time of a production-ready release there is no response in the linked gjsify issue, I will consider hosting the related [types](https://github.com/schnz/gTile/tree/types) repository under the gTile organization and publish the packages on the public NPM registry.

**Comment by kamil-karkus** [comment]
@schzn so far so good, I know it's not the beta yet but looks promising on ubuntu 23.10. Great job!

**Comment by SoiledBrush** [comment]
Settings won't load after last commit. 
The settings of extension gTile@vibou had an error:
```
Error: GSettings key animation not found in schema org.gnome.shell.extensions.gtile
```

Stack trace:
```
_checkKey@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:779:23
createCheckedMethod/<@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:729:30
#buildGeneralPage@file:///home/user/.local/share/gnome-shell/extensions/gTile@vibou/prefs.js:115:24
fillPreferencesWindow@file:///home/user/.local/share/gnome-shell/extensions/gTile@vibou/prefs.js:97:42
_loadPrefs@resource:///org/gnome/Shell/Extensions/js/extensionPrefsDialog.js:36:18
async*_init@resource:///org/gnome/Shell/Extensions/js/extensionPrefsDialog.js:21:14
ExtensionPrefsDialog@resource:///org/gnome/Shell/Extensions/js/extensionPrefsDialog.js:12:4
OpenExtensionPrefsAsync@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:139:33
async*LaunchExtensionPrefsAsync@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:126:14
_handleMethodCall@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:373:35
_wrapJSObject/<@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:408:34
_init/GLib.MainLoop.prototype.runAsync/</<@resource:///org/gnome/gjs/modules/core/overrides/GLib.js:266:34
```
But auto-close now works👍️

**Comment by schnz** [comment]
Aahh yeah, I accidentally pushed the `prefs.ts` in my first commit and haven't removed it since. There wasn't supposed to be a preferences dialog for the extension as of now. And there probably won't be one for a while. Its the least prioritized item on the bucket list.  I'll first strive to feature-completeness.

I suppose this only happens when you try to open the extension preferences, correct? Otherwise this would be a bug/error.

**Comment by SoiledBrush** [comment]
@schnz Yeah, prefs was working on previous commit. So I though it was a bug.

**Comment by schnz** [comment]
@hengestone: Thanks for the coffee(s) :heart: 

---

> So I though it was a bug.

Rightfully so. The preferences dialog that was/is present was merely a quick & dirty experiment on my side because it requires a re-implementation from scratch. It wasn't supposed to end up in the repo in the first place.


----

I just added support for a new set of hotkeys. The current hotkey support is as follows:

- [x]  Global hotkeys (<kbd>Super+Return</kbd>)
- [x] Overlay hotkeys (Arrow keys, Vi movement keys, <kbd>Shift+Left</kbd>, ...)
- [ ] Autotile hotkeys (<kbd>1</kbd>, …, <kbd>9</kbd>, <kbd>0</kbd>)
- [ ] Action hotkeys (used for window resize and movement manipulation, unbound by default)
- [x] **Hot & New:** Preset hotkeys (<kbd><Super><Control>KP_1</kbd>, …. Used to rotate through predefined preset configs)

The window resizing is still buggy with some applications and also doesn't work on fullscreen applications and otherwise maximized windows. I am aware of the issue but its also not a priority as of now (and is probably be fixed pretty quickly anyways).

**Comment by schnz** [comment]
@kamil-karkus What a pleasant surprise from you as well! Appreciate the support!

---

To everyone: Could you provide me a little insight on how you use the gTile hotkeys for window management? In particular, I am interested in how you use the autotile feature, or if you use it at all! 

Background: I am currently working on the auto tiling feature (the one with the <kbd>1</kbd>, ... <kbd>0</kbd> shortcuts and also the split into a main window and secondary windows). I am planning to introduce customizable grids for autotiling, similar on how you can define window resizing presets. The plan is to allow user-defined tiling specs according to this grammar:

```EBNF
gridspec := [ ( colspec | rowspec ) ] .
colspec  := "cols" "(" cellspec { "," cellspec } ")" .
rowspec  := "rows" "(" cellspec { "," cellspec } ")" .
cellspec := <number> [ ("d" | ":" ( colspec | rowspec ) ) ] .
```

For instance, consider the following spec:

```sql
cols(1, 1:rows(1,1), 1)
-- The numbers represent the weight of the cell relative to its siblings.
-- So an equivalent description would be
cols(3, 3:rows(2,2), 3)
```

It would create the following grid layout (where `(1, 0)` references the cell in the second column and the first row):

![grid](https://github.com/gTile/gTile/assets/3457747/5b8adb28-a3f5-43c7-973f-0fc88a6791b2)

The largest area found (or the first one, if multiple largest ones exist) would be considered the "primary" area where the focused window is being placed. The remaining windows would be placed rather arbitrarily in the remaining cells, until either no windows are left for placement or no cells are remaining for further windows.

Notice how you can provide a `d` suffix to the numbers that represent the cell weight. For instance, consider this spec:

```sql
cols(3d, 3d, 3d);
```

The `d` would declare the respective cell as `dynamic` in the sense that gTile would evenly distribute windows to dynamic cells if they would otherwise have no cell left to be placed. So in the first example above, gTile would place at most 4 windows and ignore any additional windows. Now in this case, gTile would take all windows on the respective screen and distribute them to the dynamic cells. In fact, the spec above would behave exactly as the current <kbd>3</kbd> hotkey for autotiling. But it is more flexible in that you can define areas where only a single window should be placed, and areas that "take on the rest".

So any feedback on the usefulness of this approach and whether you could incorporate that in your workflow would be nice!

@scherepanov: I am especially interested in your opinion. I know you envisioned something even more powerful with some sort of delegation to a userland script. What do you think of my proposal and what are the key aspects that you find missing there?

**Comment by ritchie888** [comment]
@schnz personally, I don't envision using an autotiling function. I use a 32:9 ultrawide monitor, and used gTile in a basic sense. I have two configurations: one where the monitor is divided into three equal parts vertically, and half horizontally, and the second where the middle third is wider and the two thirds either side are narrower (see image below). I used hotkeys exclusively, with gTile configured using the numpad keys, paired with super+ctrl/alt. I barely used the mouse select, as previously mentioned. Thanks again for all your work on this.

![gTileLayout](https://github.com/gTile/gTile/assets/1597385/8e04d5a7-ba2e-4bf2-85d3-91c3d0e4e78e)

**Comment by okoponen** [comment]
@schnz 
Not using auto-tiling, or the manually selecting grid with mouse.

[gTile-config.txt](https://github.com/gTile/gTile/files/13259456/gTile-config.txt)

I use Ubuntu.
Main flow I have set up, is moving application window manually from screen to screen via super + shift + left/right (non-gTile shortcut) or via mouse.

Then I either i use 3 sets of preset cycles, used with
* alt + super + KP_*
* ctrl + super + KP_*
* shift + super + KP_*

To me, this was the "easiest" way to setup auto-tiling-like setup, but with more manual control depending on my current need.

Comparing the need I have solved with gTile, the ubuntu 23.10 ubuntu-tiling-assistant can be configured to settle ~90% of my needs. But I still find myself missing couple of the missing gTile configs I had, mainly
* Splitting the screen either vertically or horizontally on 1/3 and 2/3 ratio
* Using the resize*5 to jave only one window, but centered on the screen ( I have 2 QHD, one on both sides of my 43" 4K, I like to sometimes use the partially covered 4K with centered data as "reading mode".

Thanks for the hard work, really appreciate it!

**Comment by Gei0r** [comment]
I also don't use auto-tiling. I only use Super+Enter and select the area with the mouse.

This already works with your current WIP, so thank you so much!

**Comment by yktoo** [comment]
I rely heavily on gTile, but I never use autotiling or mouse positioning, only hotkeys: **Ctrl+Win+<Keypad>** or **Win+Alt+<Keypad>**.

What I actually miss in the original extension is a setting to address my whole desktop, which consists of 2 screen side-by-side: 
* Ctrl+Win+<Keypad> to position the window on the left monitor
* Win+Alt+<Keypad> for the right one

**Comment by schnz** [comment]
> This already works with your current WIP, so thank you so much!

Thanks for the generous support! Also thanks to the other generous donor who hasn't actively participated yet! it raises my motivation to deliver a good production-ready extension even higher!

---

> What I actually miss in the original extension is a setting to address my whole desktop, which consists of 2 screen side-by-side:

Interesting. While technically possible to make hotkeys target only a specific monitor, I consider this out of scope at the moment. Maybe its an option for you to use the <kbd>Win+Shift+Arrow</kbd> (Gnome) hotkey to move the window to the desired monitor and then use the regular gTile hotkeys to resize the window. gTile also has a built-in hotkey to move windows from one monitor to the next (in a rotating manner). It's not yet implemented in my current preview but maybe that’s an option too.

**Comment by yktoo** [comment]
> While technically possible to make hotkeys target only a specific monitor, I consider this out of scope at the moment. Maybe its an option for you to use the Win+Shift+Arrow (Gnome) hotkey to move the window to the desired monitor and then use the regular gTile hotkeys to resize the window. gTile also has a built-in hotkey to move windows from one monitor to the next (in a rotating manner). It's not yet implemented in my current preview but maybe that’s an option too.

In my view, it's just an additional checkbox in the settings: when enabled, gTile takes the entire desktop dimensions as base for calculations instead of the active window only.

As I have 2 identical monitors, I'd then configure the tile width, for example, as ⅙ instead of ⅓.

The reason I want this is I'd use a single keystroke to position a window, instead of `Win+Shift+Arrow` and then position. But I agree, it's something the original extension wasn't able to do either, so perhaps out of scope for now.

**Comment by schnz** [comment]
Ohhh now I see what you mean. I misunderstood you the first time around. Well I suppose your suggestion has its pitfalls. The way I understand that Gnome calculates the work area (the virtual combined area of all monitors) is by essentially drawing a big rectangle around all the monitors. But not everything inside that area is necessarily an area where the user can see the content.

For instance: take a look at my setup:
![image](https://github.com/gTile/gTile/assets/3457747/51e15d85-2a68-4075-b874-0b2a044f7235)

That option wouldn't make any sense in that environment. But I can see that it can be useful for a setup such as yours. I'll make a note in my backlog about that feature. I'll maybe consider implementing it (but the implementation would be a bit time-consuming, so nothing I'll consider for the public release).

**Comment by scherepanov** [comment]
gTile very nicely support multiple monitors.
I have horizontal 43'' and two vertical 27'' on sides.
Built-in Gnome shortcut Super-Shitf-Arrow moves window to adjacent monitor in direction of arrow.
Next, use gTile keyboard control (my preference) or keyboard shortcut to place window in desired position.
Works just fine and very intuitive, with cost of additional keyboard hit.
I do not think support of immediate window placement on multiple monitors should be part of basic functionality.
Though, users should be able to add whatever window placement with help of custom plugins.... I hope we will be able to implement it.
Observing user requests for gTile functionality for many years make me think it is not possible to satisfy every users. Some people wants stuff that I think nobody would ever consider to use. That does not mean that some user requests should not to be implemented.
What we need to do - define basic functionality of gTile, and support only basic functionality.
Support plugins. Let user implement all exotic requests in plugins by themselves. Plugins should be really easy to write and implement. 
Separation of gTile functionality into fully supported basic and user-written plugins would cover by far tiling and windows placement. At same time simplify support.
I think I have a pretty good idea how we can do plugins. We will return to plugins when gTile will be supported on gnome 45, that is a prime objective.
About user request - it is outside of scope of this project, and seems to be have a very niche audience. That does not mean it should not be implemented as plugin, preferably by user himself.

**Comment by scherepanov** [comment]
Finally got Ubuntu 23.4. New gTile mostly works! Great job @schnz !

**Comment by schnz** [comment]
> Finally got Ubuntu 23.4. New gTile mostly works! Great job @schnz !

Glad it works out for you. I just commited the autotiling feature. We are nearing feature parity with the legacy version! :muscle: 

From the top of my head whats missing for the beta test:
- [] action hotkeys (global hotkeys for direct window manipulation)
- [] move to next monitor hotkey
- [] snap to neighbors (which I intend to implement a little different and call it "autogrow")
- [] few cosmetic changes to the overlay and wiring the two action buttons for autotiling

After that it should be pretty much complete. The preference dialog is something that will require a little more effort. Yet I don't think its critical to have for the beta test.

**Comment by pakwin9133** [comment]
> Finally got Ubuntu 23.4. New gTile mostly works! Great job @schnz !
Hi @scherepanov , may I know how to compile and install the latest commit which is branch "Next" ?

**Comment by peter9teufel** [comment]
Tried to build and install the latest version from the "next" branch on Ubuntu 23.10 according to the installation instructions in #341 and get no errors so far, but "gnome-extensions list" does not list the gTile extension to be installed afterwards. Any suggestions?

Thanks for your efforts and best regards,
Peter

**Comment by kamil-karkus** [comment]
> Tried to build and install the latest version from the "next" branch on Ubuntu 23.10 according to the installation instructions in #341 and get no errors so far, but "gnome-extensions list" does not list the gTile extension to be installed afterwards. Any suggestions?
> 
> 
> 
> Thanks for your efforts and best regards,
> 
> Peter

Did you re-login?

**Comment by peter9teufel** [comment]
> > Tried to build and install the latest version from the "next" branch on Ubuntu 23.10 according to the installation instructions in #341 and get no errors so far, but "gnome-extensions list" does not list the gTile extension to be installed afterwards. Any suggestions?
> > Thanks for your efforts and best regards,
> > Peter
> 
> Did you re-login?

Hi, thanks for the fast response. Had another long build job running, after logging out and login afterwards it is working fine, thanks!

**Comment by scherepanov** [comment]
> > Finally got Ubuntu 23.4. New gTile mostly works! Great job @schnz !
> > Hi @scherepanov , may I know how to compile and install the latest commit which is branch "Next" ?

PR #341 comment has instructions how to build and install gTile from next branch.

**Comment by schnz** [comment]
### 🎉🎉 **Ready for Beta!** 🥳🥳

It took a bit longer than expected but the rewrite has finally reached feature parity and I've just pushed the latest commit!

Please check #341 for the current working state and read the latest commit message for the added features and bug fixes. 

---

**I plan the beta phase to be approximately 1 week**. The plan is to merge the PR after that and I will suggest to @scherepanov  to release that version on gnome.extensions.org.

During this time, I ask for your help! Please test the current version for anything that meets the eye. That could be bugs, performance problems, bad user experience, missing features, … I will try to keep an up-to-date list of known bugs in #341, so check the first post in the PR if a bug is already known.

Thanks!

---

One last note: Some features are not yet accessible because they are exclusively available through hotkeys

- There is no preferences dialog! (at least none that allows you to set hotkeys).
- You can edit the `org.gnome.shell.extensions.gtile.gschema.xml` file prior to building the extension to add/edit the shortcuts
- If you do that, also ensure that the defaults in the XML are not overridden through your dconf settings ( `dconf dump /org/gnome/shell/extensions/gtile/`)

**Comment by hackoder** [comment]
Nice work @schnz. Works really well!

The only bug I've encountered is that gVim doesn't tile correctly. Regardless of keyboard or gui tiling, gVim's window is _sized_ correctly, but doesn't move to the correct position on the screen. Essentially, it keeps its top/left coordinates, but gets sized correctly. No other app is doing this.

**Comment by scherepanov** [comment]
Great! Will follow you plan - beta for a week, then merge PR and release to EGO.

**Comment by okoponen** [comment]
Thanks @schnz !
I installed

**Comment by SoiledBrush** [comment]
Can confirm the issue @okoponen defined.
Have to activate gTile with Super+Enter after logging in or after waking from sleep for shortcuts to start working.

**Comment by schnz** [comment]
Thanks for the reports! 

@hackoder : I can confirm the issue. Will work on it.

@okoponen  @SoiledBrush : Confirmed. I was indeed first registering the shortcuts the first time the overlay was opened. Fixed in the latest version.

**Comment by scherepanov** [comment]
Yes, sometimes gnome-shell is not moving window for no apparent reason. Seems to be a bug in some underlying API. Yes, we had to always do move_frame and then move_resize_frame to make it working for all windows.

**Comment by schnz** [comment]
@marinierb That does indeed sound like some kind of phantom window. Can you think of any application running in the background that might be detected by gTile as such? Are there any more information you can provide to reproduce the bug? Maybe a (scaled down) screenshot as well?

**Comment by marinierb** [comment]
@schnz Hah! This is the culprit: https://extensions.gnome.org/extension/2087/desktop-icons-ng-ding/
The problem goes away when I disable it.

**Comment by marinierb** [comment]
Everything else works well for me. Thanks for doing this!!

**Comment by schnz** [comment]
Glad you found the root cause. I'll take a look at the extension to see whether gTile can detect it.

I also _just_ pushed a commit that fixes another problem with autotiling (the feature you refer to) on multiple monitors. It was rearanging windows from all monitors of the current workspace as opposed to only the windows on a single monitor.

**Comment by JMurph2015** [comment]
Sorry I haven't been around much recently (a lot of other code to work on :grin: ), but I just wanted to toss in that I've wanted to create an abstraction layer between gTile and the Gnome GJS API for a while.  Basically restructuring the code so that there's a main controller class that works entirely in abstract terms and then have it call an API abstraction layer to move/resize windows/register hotkeys/etc.  That way if Gnome changes underneath us, we just have to update the abstraction class and all feature development can go into the controller class with no baggage from Gnome.

**Comment by tun0** [comment]
I noticed one regression: when a window is maximized, resizing doesn't work. Un-maximizing the window makes the resizing work again.

Also, I always had issues with resizing `urxvt`: it'd leave gaps at the right & bottom due to `urxvt` own snap-to-font-size-during-resize feature (no idea how to properly call it :sweat_smile:). With the beta it doesn't seem to work at all. The "workaround" still works though (goal: have it use it use the right two-thirds of my screen): use `win`-`right` to snap it to right half of the screen, then drag the left side to towards any window already occupying the left one-third of the screen.

**Comment by schnz** [comment]
@tun0 Regarding the issue with maximized windows that cannot be resized: Which version of the extension are you using? This should have been fixed prior to the beta (in commit hash a12c552).

I also tried to reproduce the issue with urxvt to no avail. Both moving and resizing to all screen edges works fine. I am using the `extra/rxvt-unicode` Arch Linux package though. I suppose you are on a different distro?

**Comment by tun0** [comment]
> @tun0 Regarding the issue with maximized windows that cannot be resized: Which version of the extension are you using? This should have been fixed prior to the beta (in commit hash [a12c552](https://github.com/gTile/gTile/commit/a12c55238e5d2749dcc1bb0d8de835ddba1990c1)).

I was indeed missing a bunch of updates. My bad!

> I also tried to reproduce the issue with urxvt to no avail. Both moving and resizing to all screen edges works fine. I am using the `extra/rxvt-unicode` Arch Linux package though. I suppose you are on a different distro?

I'm using the same package, but on Manjaro.

Notice the subtle difference between these screenshots:

Resized with gTille:
![Screenshot from 2023-11-10 09-27-30](https://github.com/gTile/gTile/assets/5331085/faab4b8d-53d5-4890-999b-eff1ce124d49)

Resized using "workaround":
![Screenshot from 2023-11-10 09-28-38](https://github.com/gTile/gTile/assets/5331085/c6ec4850-fe16-4b1f-b730-9f75542b6931)

The gaps do seem much smaller than before though. Than again, I got used to not use gTile to resize `urxvt`, so been a while since I saw the previous gap-size :wink:

**Comment by schnz** [comment]
@tun0: I am afraid that there is nothing I or the extension can do about that. As you already pointed out, the terminal grows in height proportional to the height of a terminal row. So pixel-perfect resizing is inherently impossible, i.e. unless you maximize the window in the vertical direction (what you essentially do with the <kbd>Super+Right</kbd> shortcut). For instance, you won't be able to have the terminal in the center of the screen (without touching the left or right edge of the monitor) while at the same time take the full height of the work area.

**Comment by tun0** [comment]
Figured as much. More of a Gnome restriction than gTile issue I guess. I'm already doing the workaround on auto-pilot, so it's not that much of an issue any way :smile:

**Comment by schnz** [comment]
@scherepanov Do you mind if I create one (maybe two) new repositories under the gTile organization for the npm package dependencies? I had hoped that the [gjsify](https://github.com/gjsify/gnome-shell/issues/5) owner would be interested in publishing an up-to-date package but I haven't heard back yet and we need publicly available npm packages to allow building the extension without a personal GitHub access token. I wouldn't yet bother to create a GitHub pipeline for publishing the packages but I would register an npm organization and invite you (+ possible others from the GitHub gTile organization?) to it, if you give me your name.

**Comment by luksa** [comment]
@schnz Thanks so much for this rewrite. I've just tested the new insets stuff, which seems to work great, but the `window-margin` setting is not yet honored.

**Comment by luksa** [comment]
@schnz I added support for `window-margin` here: https://github.com/luksa/gTile/commit/fbc3e3c4e4c5fc6d1b8706a56dd9fe5db2ee40e4

Feel free to use it (and make any required corrections).

**Comment by kamil-karkus** [comment]
@schnz I've just tested latest version on Ubuntu 23.10 and few things comes to my mind but I can't test it on 23.04 right now.
1. When I activate super + enter the grid doesn't close automatically. 
2. Phantom window on clean ubuntu 23.10 when I press super + enter right after unlocking. You can also reproduce this after minimize all the apps. 
3. I'm almost sure I could use modifier like alt or shift for super + enter to increase/decrease windows size.

**Comment by kamil-karkus** [comment]
@schnz I don't know if it's a bug or not but overlay window isn't rendered correctly, but this is not a regression, since same thing happens on gnome 44 and below 
https://i.imgur.com/6MkiAfd.png

**Comment by schnz** [comment]
### Beta Completed

Thanks to everyone who participated during the beta phase.  Many bugs that I was unaware of were fixed during that time. I just committed what I consider the final state of the rewrite. I believe its ready to be published. I apologize that the preference dialog is not part of the first public release. There is a section in the README for the time being that explains how to change settings with the DConf editor. Its less convenient but it is a workaround until the preference dialog is rewritten too (which I also plan to do but in a separate PR).

Also thanks for all the support, both verbally and  even monetary from some of you! I really appreciate that and it helped me to stay motivated during the rewrite.

@scherepanov I removed the `draft` flag from the PR. Since I consider you the core maintainer of gTile I would like for you to have a final look at the PR and if you think it requires any changes prior to being merged.  Should you be happy with its current state, feel free to merge it at your own discretion. In the most recent commit i added a GitHub workflow that is supposed to automatically create a release whenever a tagged commit is created. But I am unsure if that can be tested. I believe it has to be tested once the PR is merged.

---

> @schnz I don't know if it's a bug or not but overlay window isn't rendered correctly (look at font in the title and buttons border), but this is not a regression, since same thing happens on gnome 44 and below
https://i.imgur.com/6MkiAfd.png

I also noticed that. Yet I couldn't figure out why this is the case. I also believe its not a regression because I initialize the GUI elements in the same way as the original extension did.

**Comment by marinierb** [comment]
The **"phantom window"** problem I was having is fixed! Thank you!

**Comment by tun0** [comment]
> As you already pointed out, the terminal grows in height one terminal row at a time. So pixel-perfect resizing is inherently impossible, i.e. unless you maximize the window in the vertical direction

Could it be viable to have `gTile` do the maximizing vertically thing, obviously this would only apply for windows aligned to the left/right edge of a screen.

**Comment by schnz** [comment]
> Could it be viable to have gTile do the maximizing vertically thing, obviously this would only apply for windows aligned to the left/right edge of a screen.

In my opinion: no. For a couple of reasons:

* The window _would_ take the entire height of the work area, wouldn't it be for the applications itself that denies that. In my opinion, there should be an option on the application-level to opt-out from that behavior.
* A few aspects why I wouldn't  wanted this to be implementing in gTile:
  * If the user configured an inset larger than 0, it would practically (yet implicitly) prevent to have vertical fullscreen windows. So the behavior would be different depending on whether an inset is configured or not - which is not intuitive.
  * To be consistent, this feature would have have to be applied for all gTile operations, not just the "window placement through grid selection"-feature. There are also shortcuts to *expand* window edges towards an arbitrary direction. There are shortcuts to *move* windows towards an arbitrary direction. And some more (such as autogrow). All of these features (IMO) should be expected to put a window to full screen as soon as it touches the edge of the screen and takes the full height of the work area. IMO that would, again, be an inconsistent behavior. Why would a window become maximized when I merely want to move it?

May I suggest to consider using a different terminal emulator? I can recommend alacritty.
